src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import order.upper_lower
+
+/-!
+# Young diagrams
+
+A Young diagram is a finite set of up-left justified boxes:
+
+```text
+□□□□□
+□□□
+□□□
+□
+```
+This Young diagram corresponds to the [5, 3, 3, 1] partition of 12.
+
+We represent it as a lower set in `ℕ × ℕ` in the product partial order. We write `(i, j) ∈ μ`
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main functions:
+
+  - `young_diagram` : basic definition
+  - `young_diagram.card` : the cardinality (number of boxes)
+
+  - various constructors:
+      `young_diagram.has_emptyc` : the empty Young diagram is (∅ : young_diagram)
+
+## Notation:
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+section young_diagram
+
+/-- A Young diagram is a lower set in ℕ × ℕ -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+instance young_diagram.set_like : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := λ μ, (μ.cells : set (ℕ × ℕ)),
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma young_diagram.mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). -/
+lemma young_diagram.nw_of (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
+  (hi : i1 ≤ i2) (hj : j1 ≤ j2) (hcell : (i2, j2) ∈ μ) : (i1, j1) ∈ μ :=
+μ.is_lower_set (prod.mk_le_mk.mpr ⟨hi, hj⟩) hcell
+
+instance young_diagram.has_subset : has_subset young_diagram :=
+{ subset := λ μ ν, (μ : set (ℕ × ℕ)) ⊆ ν }
+
+/-- Cardinality of a Young diagram -/
+def young_diagram.card (μ : young_diagram) : ℕ := μ.cells.card
+lemma young_diagram.card_def (μ : young_diagram) : μ.card = μ.cells.card := rfl
+
+section μ_empty
+/-! The empty Young diagram -/
+
+/-- The empty Young diagram is (∅ : young_diagram). -/
+instance young_diagram.has_emptyc : has_emptyc young_diagram :=
```

Comment:
Presumably young diagrams form a lattice? If so, I'd lean towards using `bot` and `le` instead of `empty ` and `subset`. That lattice structure ought to be straightforward, so I think can go in this PR

Reply:
(via `distrib_lattice.lift` or similar, cc @YaelDillies who will likely know the exact name)

Reply:
Yes, [Young's lattice](https://en.wikipedia.org/wiki/Young%27s_lattice). The main thing I know about it is that the number of paths from the bottom to a particular diagram gives the dimension of that particular irreducible representation of the symmetric group. (More generally, the number of paths from a diagram up to a higher diagram is its multiplicity when inducing.)

Reply:
[`function.injective.distrib_lattice`](https://leanprover-community.github.io/mathlib_docs/order/lattice.html#function.injective.distrib_lattice)

Reply:
Thanks @YaelDillies, I was thrown off by #10704

Reply:
I’m open to this (@YaelDillies has made a similar suggestion on the Zulip re: implementing `lower_finset` which would be the generalization of this lattice structure). But I’m a little wary because there are other lattice structures on integer partitions and young diagrams, such as dominance order (https://en.m.wikipedia.org/wiki/Dominance_order). This situation may arise for other combinatorial objects too, e.g. on permutations there’s strong and weak Bruhat order.

Reply:
Usually what happens in those cases is that we make type synonyms. For example, the colexicographic order is a common order on `finset ℕ`, but that doesn't mean we don't order `finset α` by inclusion! Instead, we put the colexicographic order on [`finset.colex α`](https://leanprover-community.github.io/mathlib_docs/combinatorics/colex.html#finset.colex), a type synonym.

Reply:
OK, I have added the `distrib_lattice` instance and changed `empty` and `subset` to `bot` and `le`. I updated the docstring accordingly.

@YaelDillies I looked at `order/upper_lower.lean` but it seems like writing `lower_finset` would be a much larger PR, the API looks very extensive there. So, for now I have just written it for `young_diagram`. I'd be happy to do a limited `lower_finset` implementation but I would need clear instructions.